### DISCUSSION TITLE
Performance notes (windows)

### DISCUSSION THREAD

**Discussion by fredemmott** [qpdf/qpdf#715] [General]
Just some rough notes on bottlenecks for now

In [OpenKneeboard](github.com/fredemmott/openkneeboard), I:
1. load the PDF file
2. extract the outlines so I have a "TOC entry text" -> page number map
3. extract all internal hyperlinks in the document, so I have page number -> rect -> destination page

I'm using VS2022's performance profiler.

Will post separate comments for each point as I get to them

**Comment by fredemmott** [comment]
# GetAllPages(), Tokenizer(), tell

This is the biggest single thing in the profiler output (sorry, visual studio doesn't seem to have export functionality). This is mostly inside the MS C runtime library.

If I use `qpdf.processFile()` on https://www.digitalcombatsimulator.com/en/files/3318384/, steps 1-2 take 8s, 17s for the whole thing (1-3).

Switching to `CreateFileW()` + `CreateFileMappingW()` + `MapViewOfFile()` + `qpdf.processMemoryFile()` takes this to 3.5s for 1-2, 12.5s for everything

**Comment by fredemmott** [comment]
Tangent: profiling on linux

I don't have software using QPDF that can build on Linux; however, I've generally found `perf record` + `perf report` from the linux perf tools useful for this kind of thing when I can run it on linux

**Comment by fredemmott** [comment]
# QUtil::int_to_string

After switching to an mmapped file, this is 13% of CPU time for my case:

![image](https://user-images.githubusercontent.com/360927/173274613-0b68fb4d-ec62-4778-86a2-348a4e1f1abd.png)

While base 10 is special-cased to `std::to_string`, there's better alternatives - if a C++11 requirement is fine, consider https://github.com/fmtlib/fmt for the common bases - `fmt::format_int()` for base 10 specifically. Their latest stats are outdated, but with fmt 5.2, they claimed 4x speedup for base 10 over std::to_string

**Comment by fredemmott** [comment]
# std::string

24% of CPU is spent in the basic_string destructor:

![image](https://user-images.githubusercontent.com/360927/173275289-82b297c7-80c5-40a1-96dc-d333f16789d3.png)

... and another 25% in various forms of the constructor:

![image](https://user-images.githubusercontent.com/360927/173275732-23c5103a-71a4-484c-a49c-8a713b6427c4.png)

Naively: is something like std::string_view usable here instead to reduce copies? As of C++11, COW is not permitted for std::string, so lots of copies.

operator+ for strings is 5%, in `QPDFObjectHandles::setObjectDescriptionFromInput`; again naively, ideally this could be done lazily, otherwise, a builder or `fmt::format()` could really help here - and also avoid the int_to_string cost of that function.

**Comment by jberkenbilt** [comment]
This is extremely useful input. I will study it carefully before qpdf 11. I am aware of all the issues you mentioned, but I hadn't taken the time to really study them. Almost all this code predates my allowing C++11 (currently I have C++14 as a minimum requirement), so there may be a a number of optimizations that are available to me now that weren't previously available.

I would expect the tokenizer to dominate. I believe I just mentioned in another ticket recently that the tokenizer is "pretty good" but could be improved by reducing the amount of string copying.

I'll look at fmt, but I will probably not use it. A significant feature of qpdf is how few external dependencies it has. I have to weigh this against performance. There have also been some requests for me to use a more efficient zlib implementation. Maybe there are some ideas I can take from it since I would need only a fraction of its functionality.

I'm glad you have a way to use an mmapped file on Windows. A while ago, I tried the mmap -> read memory approach on Linux and found it to be very unstable. After considerable research, I found lots of resources about why it's unstable and abandoned that path. I have notes about it somewhere. My guess is that you'd get more gain on Windows from that anyway. While it definitely can help in a UNIX type environment, Linux is already pretty good about buffering files.

**Reply by fredemmott** [reply]
> I'll look at fmt, but I will probably not use it.  A significant feature of qpdf is how few external dependencies it has

Maybe later then :) if/when C++20 is a practical choice for the project's need, `std::format` implementations are usually derived from fmt

**Reply by fredemmott** [reply]
> I'm glad you have a way to use an mmapped file on Windows.

In case it's useful to others, here's my windows-specific implementation: https://github.com/fredemmott/OpenKneeboard/blob/8e78e816f53df802708a19fcfb11faf91c0e9ea8/src/app/app-common/tabs/PDFTab.cpp#L251-L282

> My guess is that you'd get more gain on Windows from that anyway. While it definitely can help in a UNIX type environment, Linux is already pretty good about buffering files.

This definitely matches my experience :)

It might be worth doing some custom buffering within the file path - e.g. do all OS-level reads in page size (usually 4KB) instead of single-byte stdlib ops

**Reply by jberkenbilt** [reply]
> It might be worth doing some custom buffering within the file path - e.g. do all OS-level reads in page size (usually 4KB) instead of single-byte stdlib ops

I'm delegating that to stdio or the std::ostream equivalent. :-)

**Comment by jberkenbilt** [comment]
It's possible that this may slip past 11.0. qpdf 11 is getting to have lots of changes in it, and I don't want to put it off too much farther, but if there are some quick wins or some optimizations I can make in critical sections of code like the tokenizer, that would be a big win. I have an issue at work right now that would benefit greatly from improving qpdf's tokenizer. A while ago I actually created a flex-based tokenizer but abandoned it because the performance wasn't that much better and making it work for arbitrary strings required lots of hacks. A few times I have tried to go through and see if I could squeeze more performance out of it, but not recently.

**Comment by jberkenbilt** [comment]
Some of the items discussed here, particularly improvements to the tokenizer, are in qpdf 11, which I hope to get out within the next few days (still waiting for one change). I haven't addressed int_to_string at this time.

**Reply by fredemmott** [reply]
Thanks; on the attached PDF:

| | 10.6.3.0cmake1 | 11.1.0 |
| --- | --- | --- |
| Steps 1-2  | 5180ms | 4815ms |
| Steps 1-3 | 8491ms | 7697ms |

[DCS_ReferenceGuide_v28C.pdf](https://github.com/qpdf/qpdf/files/9579856/DCS_ReferenceGuide_v28C.pdf)

That's a nice improvement :)

**Reply by jberkenbilt** [reply]
Mostly thanks to @m-holger

**Reply by fredemmott** [reply]
For comparison, when using pdfium on the same file:

Steps 1-2: 95ms
Steps 1-3: 679ms

Edit: that's not like-for-like, that's with release pdfium and debug qpdf. That (accidentally) worked with pdfium in my debug builds as pdfium's API is C, so doesn't have the C++ standard library compatibility issues that exist on windows between different kinds of builds

QPDF relwithdebinfo numbers are 328 and 727 respectively

**Reply by jberkenbilt** [reply]
A lot more resources go into pdfium, and it's a full PDF interpreter, so I don't even pretend to be in competition with it. I have sometimes cited it as the thing that may eventually make qpdf irrelevant. That said, it looks like qpdf's release version is not that far behind it for steps 1-3. I have been told that qpdf is faster than ghostscript, but I don't know if that's still the case with >= 9.56 and the new PDF interpreter written in C. Anyway, those are good lower bounds. While I do want qpdf to be efficient, I have historically been willing to trade a little in the efficiency department in favor of maintainability. That said, I'd say string manipulation would be the next frontier for optimization in qpdf now that the parser and tokenizer have gotten some attention. If C++ 20 gives us out-of-the-box solutions, maybe we just wait for it. When the oldest LTS versions of popular Linux distrbutions have C++ compilers that support C++-17. I may be willing to bump up to that as the minimum standard. It's pretty important for me to keep qpdf fairly easy to build using standard toolchains and as few external dependencies as possible. For now, that still means C++-14.

**Reply by fredemmott** [reply]
Yeah, I'm sticking with qpdf for pretty much the reasons you mentioned, and there's not enough of a win there for me to justify pdfium's multi-hour build time (including dependencies that need to be rebuilt with clang's clang and libc++ on windows) or depending on a pre-built blob in my build.

There's 3 big things I'm seeing scope for:
- string copies: `std::string_view` or equivalent: included in C++17, but pretty straightforward to do a minimal custom version that'd be sufficient
- string-to-int conversions: fmtlib does this much better and is compatible with C++11; it seems likely that most standard libraries will use fmt to implement `std::format` in C++20. Importing their string-to-int implementation (without the format string support) would be relatively small if desired, to avoid the extra dependency
- lazy-loading/evaluation/conversion; for example:
  - `QPDFObjectHandles::setObjectDescriptionFromInput`
  - encoding conversions: I **greatly** appreciate I don't have to worry about which elements are 'PDFDocEncoding', which are 7-bit ASCII, which are UTF-16BE, though (from an admittedly very brief look that could easily be wrong) it looks like (most?) things are eagerly converted to UTF-8?

**Reply by m-holger** [reply]
@fredemmott Do you happen to have corresponding timings for 11.3.0? Would be interesting to see.

**Comment by jberkenbilt** [comment]
In #781 @m-holger eliminated almost all calls to `QUtil::int_to_string` and `QUtil::uint_to_string`, replacing them with `std::to_string`. This should help and will be in 11.1.1 (or 11.2, whatever is next) when I release it. I have at least one more minor issue to resolve and will give it a few weeks (ish) to let other issues trickle in before I release.

**Comment by jberkenbilt** [comment]
Another user has reported odd performance issues on Windows in #798. I can see that there would be value in doing a more in-depth study of qpdf's performance on Windows and what we can do about it.

**Comment by jberkenbilt** [comment]
I'll be releasing 11.1.1 later today. Linux performance benchmarks show close to 5% improvement. A lot of that is elimination of almost all calls to `int_to_string`, replacing with `std::to_string`. `int_to_string` was already using this internally in most cases.
